corehq/apps/translations/views.py

```diff
@@ -0,0 +1,121 @@
+    page_title = _('Convert Translations')
+    urlname = 'convert_translations'
+    template_name = 'convert_translations.html'
+    section_name = ugettext_noop("Translations")
+
+    @property
+    @memoized
+    def convert_translation_form(self):
+        if self.request.POST:
+            return ConvertTranslationsForm(self.domain, self.request.POST, self.request.FILES)
+        else:
+            return ConvertTranslationsForm(self.domain)
+
+    @property
+    @memoized
+    def _uploaded_file_name(self):
+        uploaded_file = self.convert_translation_form.cleaned_data.get('upload_file')
+        return uploaded_file.name
+
+    @staticmethod
+    def _parse_excel_sheet(worksheet):
+        rows = [row for row in worksheet.iter_rows()]
+        headers = [cell.value for cell in rows[0]]
+        source = headers.index('source')
+        translation = headers.index('translation')
+        occurrence = headers.index('occurrence') if 'occurrence' in headers else None
+        context = headers.index('context') if 'context' in headers else None
+        return rows, source, translation, occurrence, context
+
+    def _generate_translations_for_po(self, worksheet):
+        rows, source, translation, occurrence, context = self._parse_excel_sheet(worksheet)
+        translations = {worksheet.title: []}
+        for index, row in enumerate(rows[1:]):
+            _occurrence = row[occurrence].value if occurrence else ''
+            _context = row[context].value if context else ''
+            translations[worksheet.title].append(
+                Translation(
+                    row[source].value,
+                    row[translation].value,
+                    [(_occurrence, None)],
+                    _context)
+            )
+        return translations
+
+    def _generate_po_file(self, worksheet):
+        translations = self._generate_translations_for_po(worksheet)
+        po_file_generator = PoFileGenerator()
+        po_file_generator.generate_translation_files(translations, {})
+        return po_file_generator
+
+    def _generate_excel_file(self):
+        uploaded_file = self.convert_translation_form.cleaned_data.get('upload_file')
+        po_file = polib.pofile(uploaded_file.read())
+        wb = openpyxl.Workbook()
+        ws = wb.worksheets[0]
+        ws.title = "Translations"
+        ws.append(['source', 'translation', 'context', 'occurrence'])
+        for po_entry in po_file:
+            ws.append([po_entry.msgid, po_entry.msgstr, po_entry.msgctxt,
+                       po_entry.occurrences[0][0] if po_entry.occurrences else ''])
+        return wb
+
+    def _po_file_response(self):
+        uploaded_file = self.convert_translation_form.cleaned_data.get('upload_file')
+        worksheet = openpyxl.load_workbook(uploaded_file).worksheets[0]
+        po_file_generator = self._generate_po_file(worksheet)
+        content = open(po_file_generator.generated_files[0][1], 'r', encoding="utf-8").read()
```

Comment:
i thought since its not stored to an object, there wont be any open references left?

Reply:
I don't know the answer to that - I know that `with open()` is fine and standard practice, but I'm not certain about `open()`.  I think it's one of those things that depends on the architecture and the python interpreter being used, as opposed to `with`, which does it as part of the language specification.  That is, it probably does do what you want, but it still feels weird to me.  Not that big a deal though - I think it's a style thing that you can decide.

Reply:
`with open()` is more explicit in terms of resource cleanup, and therefore is better. Using `open()` without explicitly closing the file means that the file reference hangs around as long as there is a name referencing it in some scope or garbage collection hasn't been run on the names that were referencing it. On CPython, garbage collection is somewhat deterministic because of the way reference counting is implemented, so in practice using bare `open()` will most likely not hurt anything—although it still could if the code was written in a way that hit max open file count or something like that. Most other versions of Python do not have the same guarantees with respect to garbage collection.

It is a best practice to always do explicit resource cleanup (use `with open()`) because that makes it easier to reason about the resources your code is using.

Reply:
Noted. thank you 👍